### PR TITLE
Piece definer line numbers

### PR DESCRIPTION
**Pull request by riverwanderer**
Complements search line cross-reference feature.

### CODE REVIEW

File: vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java [hunk 1]
```diff
@@ -141,7 +141,7 @@ public class PieceDefiner extends JPanel {
   private JLabel scaleLabel;
   private final Prefs prefs;
 
-  private String prototypeName; // If we're editing a prototype definition, this holds the name of it
+  private static String prototypeName; // If we're editing a prototype definition, this holds the name of it
```
Comment (BrentEaston) [inline vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java:144]: Why change this to static? It's possible to have multiple Piece Definers open at the same time, If you open a piece, then a prototype before closing the piece, the line numbers in the piece start acting like a prototype

File: vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java [hunk 2]
```diff
@@ -1280,13 +1280,24 @@ public Component getListCellRendererComponent(
...
+      else {
+        lineNumber = (index == 0 ? "" : " ".repeat(index < 10 ? 2 : index < 100 ? 1 : 0) + index + ". ");
+      }
       super.getListCellRendererComponent(list, "", index, selected, hasFocus);
+
       if (value instanceof EditablePiece) {
-        setText(((EditablePiece) value).getDescription());
+        final String valueType = ((EditablePiece) value).getType();
+
+        setText((valueType.startsWith("cmt;") ? "" : lineNumber) + ((EditablePiece) value).getDescription());
```
Comment (BrentEaston) [inline vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java:1296]: Try this:

        final String description = ((EditablePiece) value).getDescription();
        if (valueType.startsWith("cmt;")) {
          setText("<html>" + lineNumber.replace(" ", "&nbsp;") + description.substring(6));
        }
        else {
          setText(lineNumber + description);
        }


Actually, a better solution is to remove the HTML from Comment.getDescription (just have it return comment) and add it here in the renderer. The HTML is just for the Piece Definer, so the piece Definer should add it in. That will make this code a little simpler and things like Search that also call getDescription will not see the html.

File: vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java [hunk 3]
```diff
@@ -1280,13 +1280,24 @@ public Component getListCellRendererComponent(
...
       // It is incredibly inefficient for GamePieces and is not needed
       // since we overwrite the label text anyway.
+      String lineNumber;
+      if (PieceDefiner.prototypeName.isEmpty()) {
+        final int index1 = index + 1;
+        lineNumber = " ".repeat(index1 < 10 ? 2 : index1 < 100 ? 1 : 0) + index1 + ". ";
+      }
+      else {
+        lineNumber = (index == 0 ? "" : " ".repeat(index < 10 ? 2 : index < 100 ? 1 : 0) + index + ". ");
+      }
```
Comment (BrentEaston) [inline vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java:1290]: Given that the font used in the PieceDefiner is variable width, this code doesn't really have any effect, it does not cause the single digit and two digit numbers to line up. I would just removed all the leading spaces and just go with left justified numbers.

**Review by BrentEaston — CHANGES_REQUESTED** [review]
Overall, I like the direction of this a lot. Just needs a few more tweaks I think

File: vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java (hunk 1 above)
Comment (riverwanderer) [reply vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java]: > Why change this to static? It's possible to have multiple Piece Definers open at the same time, If you open a piece, then a prototype before closing the piece, the line numbers in the piece start acting like a prototype

Thanks Brent. I suspected there'd be an issue but lack of Java knowledge meant this was the quickest workaround for me. For the next build I am now using an independent means of determining if a Prototype is being processed.

PS I was unable to replicate problem behaviour to see what it looked like but I've just realised I didn't use the precise sequence of opening that you indicated above.

File: vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java (hunk 2 above)
Comment (riverwanderer) [reply vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java]: Great suggestion, thanks @BrentEaston. I also took advantage of the comment handling now in the renderer to alter the formatting of a comment. See what you think in the next build.

File: .github/ISSUE_TEMPLATE/module-bug-report.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
+---
+name: Module bug report
+about: Create a report to help improve a specific module
```
Comment (BrentEaston) [inline .github/ISSUE_TEMPLATE/module-bug-report.md:3]: Are these issue templates supposed to be in this PR?

File: vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java [hunk 5]
```diff
@@ -1281,15 +1282,47 @@ public Component getListCellRendererComponent(
...
+
       if (value instanceof EditablePiece) {
-        setText(((EditablePiece) value).getDescription());
+        final EditablePiece editableValue = (EditablePiece) value;
+        final String type = editableValue.getType().split(";")[0];
+        final String valueStr = editableValue.toString();
+        final String name = valueStr.substring(valueStr.indexOf("[name=")).split("=")[1].split(",")[0]; // extract name
+
+        // Prototype name field is null; skip first (dummy) item to avoid outputting a blank line
+        if (!name.isEmpty() || index > 0) {
```
Comment (BrentEaston) [inline vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java:1293]: This doesn't work. A Basic Piece is allowed to have an empty name and the piece name is defined in a Basic Name trait. 

You will need to pass the PieceDefiner into the Renderer via a new constructor so the renderer can query the PieceDefiner if it editing a Propertype or not

In the PieceDefiner:

Public PieceDefiner() {
   ...
   r = new Renderer(this);
   ...
}


public boolean isPrototype() {
  return prototypeName.isEmpty()
}

In the Renderer:

private PieceDefiner definer;

Renderer(PieceDefiner definer) {
   this.definer = definer;
}

...

if definer.isPrototype() {
...

**Review by BrentEaston — CHANGES_REQUESTED** [review]
Looking great, Just the issue with recognizing prototypes properly.

File: .github/ISSUE_TEMPLATE/module-bug-report.md (hunk 4 above)
Comment (riverwanderer) [reply .github/ISSUE_TEMPLATE/module-bug-report.md:3]: No. Depressing how they managed to sneak in, I thought I'd got rid. On second thoughts, maybe this PR was affected before I reverted that change from my master.

File: vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java (hunk 5 above)
Comment (riverwanderer) [reply vassal-app/src/main/java/VASSAL/counters/PieceDefiner.java]: Thanks. I couldn't work that out before. Next build incorporates these changes.

**Comment by riverwanderer** [comment]
Superseded by #12861 which contains only the line-numbering commits.

**Pull request closed without merging**